R/dock_from_renv.R

```diff
@@ -0,0 +1,90 @@
+#' Create a Dockerfile from an `renv.lock` file
+#'
+#' @param lockfile Path to an `renv.lock` file.
+#' @param distro One of "focal", "bionic", "xenial", "centos7", or "centos8". See available distributions at https://hub.docker.com/r/rstudio/r-base/.
+#' @param primary_repo Default "CRAN". Indicates which repository in the `renv.lock` file to install the renv package from.
+#' @param out_dir The directory where to write the `Dockerfile`. When `NULL` no file is written. `NULL` by default.
+#' @param use_rspm Utilize the Public RStudio Package Manager for Linux binaries. Default `TRUE`.
+#'
+#' @return A R6 object of class `Dockerfile`.
+#' @details
+#'
+#' System requirements for packages are provided through RStudio Package Manager via the {pak} package. The install commands provided from pak are added as `RUN` directives within the `Dockerfile`.
+#'
+#' The R version is taken from the `renv.lock` file. Packages are installed using `renv::restore()` which ensures that the proper package version and source is used when installed.
+#'
+#' By default, `dock_from_renv()` will utilize the [public RStudio Package Manager](https://packagemanager.rstudio.com/) (PPM) to provide linux binaries. This dramatically improves install time. This is done when `use_rspm = TRUE`.  To do so, a new file `renv.lock.dock` will be created so as to not overwrite your existing `renv.lock` file. This new `renv.lock.dock` file is used to build the Docker image with PPM binaries.
+#'
+#' \donttest{
+#' dock <- dock_from_renv("renv.lock", distro = "xenial", out_dir = getwd())
+#' dock
+#' }
+#' @export
+dock_from_renv <- function(lockfile = "renv.lock", distro = "xenial", primary_repo = "CRAN",
+                           out_dir = NULL,
+                           use_rspm = TRUE) {
+
+  lock <- renv:::lockfile(lockfile)
+
+  # change to public package manager by default, if false use default repo
+  if (use_rspm) {
+    lock$repos(CRAN = glue::glue("https://packagemanager.rstudio.com/all/__linux__/{distro}/latest"))
+    # write to tempfile
+    lockfile <- paste0(lockfile, ".dock")
+    lock$write(lockfile)
+  }
+
+  # start the dockerfile
+  dock <- Dockerfile$new(gen_base_image(distro = distro, r_version = lock$version()))
+
+  distro_args <- switch(distro,
```

Comment:
I wonder how much this is required to list all distributions.  
I guess, if the url in packagemanager does not exists, then the build will fail.  
Can't we get the system requirements using `.Platform()` or with `sessionInfo()$running`.  
I did not really explored this, just wondering.  
Thank you

Reply:
You're right.  
But I was wondering if when building the container, we can run a R process that knows the current platform and create the list of system requirements accordingly. Like {actions} seems to do: https://github.com/r-lib/actions

Reply:
If the purpose is for using with r-lib GH actions I believe they have an
action which will restore for you.

https://github.com/r-lib/actions/blob/v2-branch/setup-renv/action.yaml

Reply:
This is not to be used in GH actions, but to build the correct Dockerfile.   
Do not worry, we can let that apart for the moment.

Reply:
Can you elaborate on this a bit? I do not follow. How are you determining what the "correct" Dockerfile is? Is your anticipated use case people building on Linux machines? Or do you think that a Dockerfile should be based on the machine that was used to develop? E.g. if developed on Mac, a Mac docker image should be used, windows with a windows Docker image, etc.

Reply:
I think that if you use {distro} + `pak::pkg_system_requirements()`, it will determine which OS distribution is build during `docker build`, so that you do not have to calculate it when writing the Dockerfile.  

As far as I understand, you want to create a Dockerfile that already contains lines like:
```
- apt-get install -y libxxx
```

Instead, I suggest that during the docker build, we let `pak::pkg_system_requirements()` get the list of system req and install it with a `system()` command. Instead of the above, the Dockerfile would look like:
```
- Rscript -e 'install.packages(c("distro", "pak"))'
- R - e 'req <- pak::pkg_system_requirements(<get list from description>); system(paste(req, collapse = " & "))'
```
In this case, we do not have to specifiy the os when creating the Dockerfile.

Reply:
The OS needs to be specified in the first line of the Dockerfile in the `FROM` command. This tells us which base image to start from. These are taken from https://github.com/rstudio/r-docker. This lets the user not be worried about the installation of R within the Dockerfile. As such installing sysdeps from the Dockerfile vs the R command is 6 dozen to one, half a dozen to the other. 

My knowledge of Docker is middling at best so perhaps I'm missing something fundamental here :o